spec/firebase_app_distribution_auth_client_spec.rb

```diff
@@ -0,0 +1,84 @@
+describe Fastlane::Auth::FirebaseAppDistributionAuthClient do
+  let(:auth_helper) { Class.new { extend(Fastlane::Auth::FirebaseAppDistributionAuthClient) } }
+  let(:fake_binary) { double("Binary") }
+  let(:fake_binary_contents) { double("Contents") }
+  let(:auth) { Google::Auth::ServiceAccountCredentials }
+  let(:fake_service_creds) { double("service_account_creds") }
+  let(:fake_auth_client) { double("auth_client") }
+  let(:payload) { { "access_token" => "fake_auth_token" } }
+
+  before(:each) do
+    allow(Signet::OAuth2::Client).to receive(:new)
+      .and_return(fake_auth_client)
+    allow(auth).to receive(:make_creds)
+      .with(anything)
+      .and_return(fake_service_creds)
+    allow(fake_service_creds).to receive(:fetch_access_token!)
+      .and_return(payload)
+
+    allow(fake_auth_client).to receive(:fetch_access_token!)
+    allow(fake_auth_client).to receive(:access_token)
+      .and_return("fake_auth_token")
+
+    allow(File).to receive(:open)
+      .and_return(fake_binary)
+    allow(fake_binary).to receive(:read)
+      .and_return(fake_binary_contents)
+    allow(fake_binary_contents).to receive(:key)
+      .and_return("fake_service_key")
+      
+    allow(ENV).to receive(:[])
+      .with("GOOGLE_APPLICATION_CREDENTIALS")
+      .and_return(nil)
+  end
+
+  describe '#fetch_auth_token' do
+    it 'uses service credentials for authorization when the path is passed in' do
+      expect(auth_helper.fetch_auth_token("google_service_path")).to eq("fake_auth_token")
```

Comment:
I have changed the name of the auth token for the service credentials. The only way to get that auth token is to go through the service account path and so we can be sure that the Firebase tests aren't using the service account auth token and vice versa. I would be more than happy to add some more clarity if that doesn't seem like enough to be sure of the functionality.

Reply:
That helps! One thing left is you still aren't expecting one service account path to take precedence over the other, since if you set both the path and the environment variable it wouldn't be clear which one was used. I think setting up more specific to_receive/with/and_return in your initial setup, or moving some of that setup into the individual tests, could help with that. But I think what you've got is pretty good so feel free to keep it as is.